spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/AbstractMongoQuery.java

```diff
@@ -415,4 +429,55 @@ private Object deleteAndConvertResult(Query query, MongoEntityMetadata<?> metada
 			return writeResult != null ? writeResult.getN() : 0L;
 		}
 	}
+
+	/**
+	 * @author Thomas Darimont
+	 */
+	private class CursorBackedExecution extends Execution {
+
+		@Override
+		Object execute(final Query query) {
+
+			Class<?> javaType = getQueryMethod().getEntityInformation().getJavaType();
+			QueryMapper queryMapper = operations.getQueryMapper();
+
+			return createCursorAndWrapInCloseableIterator(query, javaType, queryMapper);
+		}
+
+		@SuppressWarnings("unchecked")
+		private CloseableIterator<Object> createCursorAndWrapInCloseableIterator(final Query query,
+				final Class<?> entityType, final QueryMapper queryMapper) {
+
+			final MongoPersistentEntity<?> persistentEntity = operations.getMappingContext().getPersistentEntity(entityType);
+
+			return (CloseableIterator<Object>) operations.execute(entityType, new CollectionCallback<Object>() {
+
+				@Override
+				public Object doInCollection(DBCollection collection) throws MongoException, DataAccessException {
+
+					DBObject mappedFields = queryMapper.getMappedFields(query.getFieldsObject(), persistentEntity);
+					DBObject mappedQuery = queryMapper.getMappedObject(query.getQueryObject(), persistentEntity);
+
+					DBCursor cursor = collection.find(mappedQuery, mappedFields);
+
+					return new CloseableIterableCusorAdapter<Object>(cursor, operations.getConverter(), entityType);
+				}
+			});
+		}
+	}
+
+	/**
+	 * @author Thomas Darimont
+	 */
+	final class StreamExecution extends CursorBackedExecution {
```

Comment:
Do you think it's possible to move that into Spring Data Commons and rather turn it into a post-processing step in case the execution returns a `CloseableIterable` but effectively needs to be a `Stream` (similarly to what we already do with `Optional`s. I think this would allow us to automatically support `Stream` as return type on all stores that implement the execution into a `ClosableIterable`.